### PR TITLE
CricInfo_week3

### PR DESCRIPTION
**Pull request by gokul-bvb**

### CODE REVIEW

File: Cricketer Info/Controller/HomePageViewController/HomePageViewController+TableViewDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
+//
+//  HomePageViewController+TableViewDataSource.swift
+//  Cricketer Info
+//
+//  Created by Gokul on 23/02/23.
+//
+
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/HomePageViewController/HomePageViewController+TableViewDataSource.swift]: Delete all the commented lines 1-7.

File: Cricketer Info/Controller/HomePageViewController/HomePageViewController+TableViewDelegate.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+//
+//  Created by Gokul on 23/02/23.
+//
+
+import UIKit
+
+extension HomePageViewController: UITableViewDelegate {
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 60
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/HomePageViewController/HomePageViewController+TableViewDelegate.swift]: Can we make the height dynamic

File: Cricketer Info/Controller/HomePageViewController/HomePageViewController+TeamManagerDelegate.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+//  Cricketer Info
+//
+//  Created by Gokul on 23/02/23.
+//
+
+import Foundation
+
+extension HomePageViewController: TeamDataManagerDelegate {
+    func didUpdateTeamData(teamData: TeamResponse) {
+        DispatchQueue.main.async {
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/HomePageViewController/HomePageViewController+TeamManagerDelegate.swift]: Explore capture list, and see if the capture list can be added inside line 12.

File: Cricketer Info/Controller/HomePageViewController/HomePageViewController.swift [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+//  Created by Gokul on 05/02/23.
+//
+
+import UIKit
+
+///Properties belong to an object, whereas variables do not. A variable can be declared without having to be associated with a particular class, or other object. A property must be associated with a particular object (i.e.: a class, enum, or struct)
+
+class HomePageViewController: UIViewController {
+    ///Instance that holds all the views for HomePage
+    let homePageViews = HomePageViews()
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/HomePageViewController/HomePageViewController.swift]: Add lazy

File: Cricketer Info/Controller/HomePageViewController/HomePageViewController.swift [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    
+    func getPlayerDataForCurrentIndexPath(_ indexPath: IndexPath) -> Player? {
+        guard let teamResponse = teamResponse, let selectedTeamId = selectedTeamId else {
+            return nil
+        }
+        let playerData = teamResponse.teams[selectedTeamId].players[indexPath.row]
+        return playerData
+    }
+}
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/HomePageViewController/HomePageViewController.swift]: Introduce ViewModel to HomePageView controller

File: Cricketer Info/Controller/HomePageViewController/HomepageViewController+Constraints.swift [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        setupPickerViewConstraint()
+        setupTableViewConstraints()
+        setupActivityIndicatorConstraints()
+    }
+    
+    func setupActivityIndicatorConstraints() {
+        view.addSubview(homePageViews.activityIndicator)
+        homePageViews.activityIndicator.translatesAutoresizingMaskIntoConstraints = false
+        homePageViews.activityIndicator.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
+        homePageViews.activityIndicator.centerYAnchor.constraint(equalTo: view.centerYAnchor).isActive = true
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/HomePageViewController/HomepageViewController+Constraints.swift]: Add anchors

File: Cricketer Info/Controller/PlayerDetailViewController/PlayerDetailViewController+Constraints.swift [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+//  Cricketer Info
+//
+//  Created by Gokul on 17/02/23.
+//
+///UI kit is a framework provided by Apple in iOS development that allows developers to create user interfaces for their applications. Importing UI kit into a Swift project is necessary because it provides a set of pre-built components and tools that can be used to create visually appealing and interactive user interfaces.
+///When you import UI kit into a Swift project, you gain access to a range of pre-built user interface components such as buttons, labels, text fields, and views. These components are designed to be easily customizable and can be used to build complex user interfaces.
+///Additionally, UI kit provides a range of tools for handling user input, animations, and transitions, which can help improve the overall user experience of your application.
+///Without importing UI kit, developers would need to build all the user interface components from scratch, which would be time-consuming and error-prone. Importing UI kit streamlines the process of creating user interfaces and helps ensure that the application is consistent with the design guidelines set by Apple.
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/PlayerDetailViewController/PlayerDetailViewController+Constraints.swift]: Write on your own

File: Cricketer Info/Controller/PlayerDetailViewController/PlayerDetailViewController+Constraints.swift [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        navigationController?.navigationBar.topItem?.backBarButtonItem = backButton
+    }
+    
+    func setupPlayerImageViewConstraints() {
+        view.addSubview(playerViews.playerImageView)
+        playerViews.playerImageView.translatesAutoresizingMaskIntoConstraints = false
+        playerViews.playerImageView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 0).isActive = true
+        playerViews.playerImageView.widthAnchor.constraint(equalToConstant: 150).isActive = true
+        playerViews.playerImageView.centerXAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerXAnchor).isActive = true
+        playerViews.playerImageView.heightAnchor.constraint(equalToConstant: 150).isActive = true
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/PlayerDetailViewController/PlayerDetailViewController+Constraints.swift:26]: Add anchors

File: Cricketer Info/Controller/PlayerDetailViewController/PlayerDetailViewController+Constraints.swift [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    
+    func setupPlayerNameLabelConstraints() {
```
Comment (gokul-bvb) [inline Cricketer Info/Controller/PlayerDetailViewController/PlayerDetailViewController+Constraints.swift]: Add private in all the functions

File: Cricketer Info/Coordinator/Coordinator.swift [hunk 10]
```diff
@@ -10,5 +10,17 @@ import UIKit
...
+///         Structs cannot conform to protocol that is
+///         It refers to any instance of a class. It’s useful when you are working with reference type only. It’s equivalent to ‘id’ in Objective-C.
+///         AnyObject can represent an instance of any class type.
+///
+//Any
+///Any can represent an instance of any (class, struct, enum) type at all, including function types. In simple language we can say that Any can deal with both reference type as well as value type object.
+// Any -> AnyObject -> NSObject
+
+///Structs cannot conform to protocols of special type AnyObject because reference types can only conform to such protocols
+///While protocols marked with Any, both value types and reference types can conform to the protocol with any special type
```
Comment (gokul-bvb) [inline Cricketer Info/Coordinator/Coordinator.swift]: Make it more short with example

File: Cricketer Info/Model/Player.swift [hunk 11]
```diff
@@ -0,0 +1,15 @@
+//
+//  Player.swift
+//  Cricketer Info
+//
+//  Created by Gokul on 21/02/23.
+//
+
+import Foundation
+
+struct Player: Codable {
```
Comment (gokul-bvb) [inline Cricketer Info/Model/Player.swift]: Add only needed protocols

File: Cricketer Info/Model/PlayerDataModel.xcdatamodeld/PlayerDataModel.xcdatamodel/contents [hunk 12]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<model type="com.apple.IDECoreDataModeler.DataModel" documentVersion="1.0" lastSavedToolsVersion="21512" systemVersion="22C65" minimumToolsVersion="Automatic" sourceLanguage="Swift" userDefinedModelVersionIdentifier="">
+    <entity name="Entity" representedClassName="Entity" syncable="YES" codeGenerationType="class">
+        <attribute name="attribute" optional="YES"/>
+    </entity>
+</model>
```
Comment (gokul-bvb) [inline Cricketer Info/Model/PlayerDataModel.xcdatamodeld/PlayerDataModel.xcdatamodel/contents:6]: Why do we need this?

File: Cricketer Info/Networking/TeamDataManager.swift [hunk 13]
```diff
@@ -0,0 +1,71 @@
+//
+//  FetchTeamData.swift
+//  Cricketer Info
+//
+//  Created by Gokul on 21/02/23.
+//
+
+import Foundation
+
+final class TeamDataManager {
```
Comment (gokul-bvb) [inline Cricketer Info/Networking/TeamDataManager.swift]: This is not required

File: Cricketer Info/Networking/TeamDataManager.swift [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+///            Note
+///            The URLSession API involves many different classes that work together in a fairly complex way which may not be obvious if you read the reference documentation by itself. Before using the API, read the overview in the URL Loading System topic. The articles in the Essentials, Uploading, and Downloading sections offer examples of performing common tasks with URLSession.
+///            Your app creates one or more URLSession instances, each of which coordinates a group of related data-transfer tasks. For example, if you’re creating a web browser, your app might create one session per tab or window, or one session for interactive use and another for background downloads. Within each session, your app adds a series of tasks, each of which represents a request for a specific URL (following HTTP redirects, if necessary).
```
Comment (gokul-bvb) [inline Cricketer Info/Networking/TeamDataManager.swift]: Remove these paragraphs, give your understandings only

File: Cricketer Info/Networking/TeamDataManager.swift [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        let decoder = JSONDecoder()
+        do {
+            let decodedData = try decoder.decode(TeamResponse.self, from: teamResponse)
+            return decodedData
+        } catch {
+            print("Failed to parse data due to error: \(error)")
+            return nil
+        }
+    }
+}
```
Comment (gokul-bvb) [inline Cricketer Info/Networking/TeamDataManager.swift]: Will improve this file after VM is done.

File: Cricketer Info/Networking/TeamDataManagerDelegate.swift [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+//  Cricketer Info
+//
+//  Created by Gokul on 21/02/23.
+//
+
+///Any:
+///Any can represent an instance of any (class, struct, enum) type at all, including function types. In simple language we can say that Any can deal with both reference type as well as value type object.
+
+///AnyObject:
+///It refers to any instance of a class. It’s useful when you are working with reference type only. It’s equivalent to ‘id’ in Objective-C.
```
Comment (gokul-bvb) [inline Cricketer Info/Networking/TeamDataManagerDelegate.swift]: Write your understandings only

File: Cricketer Info/View/ViewController+Views/HomePageViews.swift [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+class HomePageViews {
+    lazy var listTableView: UITableView = {
+        var tableView = UITableView()
+        tableView.separatorStyle = .singleLine
+        tableView.showsVerticalScrollIndicator = true
+        tableView.showsHorizontalScrollIndicator = true
+        return tableView
+    }()
+    lazy var teamPickerView = UIPickerView()
+    lazy var activityIndicator = UIActivityIndicatorView()
```
Comment (gokul-bvb) [inline Cricketer Info/View/ViewController+Views/HomePageViews.swift]: Make it a closure

**Comment by Gokul-7-7** [comment]
Merge

**Pull request merged**